### PR TITLE
feat: implement positionals

### PR DESCRIPTION
**Pull request by Gozala**
Implements positions [per article](https://observablehq.com/@gozala/deterministically-biased-fractional-indexing). There is whole bunch of code but the user facing API looks like:

```js
const position = Position.insert(identifier, {
  before: positionOfTheElementBefore,
  after: positionOfTheElementAfter
})
```

Both `before` and `after` are optional. Returned `position` can be used as an attribute in the assertions.

ℹ️ This is something we will include in the synopsys agent interface.

> 😅 After reflecting on the implementation, I'm realizing that it could have being simplified if it was designed differently. Specifically implementation entangles constraints that perhaps could have being avoided at the db layer. Specifically arithmetics is concerned with [robustness principal](https://en.wikipedia.org/wiki/Robustness_principle) and accepts arbitrary bytes as input positions, while doing it's best to produced well formed positions.
>
> I think better option would have being to just tag plain attributes and positions differently at db level. That way positions that do not parse could simply be treated as attributes as opposed to positions. It would imply that code could produce an error if you're passing attributes instead of positions, but that doesn't seem unreasonable, especially if type system along with query engine treated those as different.

### CODE REVIEW

File: src/position/major.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+/**
+ * Major byte describes the size range of the minor positions. It highly
+ * recommended to use only bytes in the  'A-Za-z' character code range meaning
+ *
+ * ```
+ * [65...90,  97..122]
+ * ```
+ *
+ * Major codes like `z:65` and `a:97` denote two digit `Minor` range, codes `Y:66`
```
Comment (jsantell) [inline src/position/major.js:30]: These codes *do not* line up with their ASCII representations, IIUC -- `A:65`, not `z`, no? `b:98` and `B:66` as well.

Comment (jsantell) [reply src/position/major.js:30]: IIUC what's the difference between `A:65` and `a:97` if both signify minor patch size, this seems to have a lot of opinions on underlying encoding if "A" and "a" have some significant relationship other than its relative position

**Review by jsantell — COMMENTED** [review]
My understanding that this encoding is used to satisfy lexicographic sorting when querying for a range of values, provide sufficient resolution for repeated insertions (where a simple float-average approach would be exhausted, and also inconsistent between machines[?]), as well as sufficient compression for small-ish (< 50 items?) collections, while supporting large (10k?) collections.

The "semver" approach satisfies resolution and compression well. I believe the ASCII encoding, however, pushes more complexity to the end-user (position values are user-facing, and as query-writers, so too is this library IIUC, maybe not if always anchoring to an existing positional[?]). Compared to non-renderable encodings of say `UInt8Array`, where a "digit" can be 0-255, and can extend with additional digits using similar compression techniques, e.g. Major digit indicates how many Minor digits there are. That being said, that's a large change. Concretely, my questions/confusion:

* What are the benefits for defining custom encoding ranges?
* Why is Base52 used for Major and Base62 used for Minor/Patch?
* Ultimately as a user-provided value, I question the effectiveness of "recommended" values to use: values should either be supported, or not, and a case where we're requiring end-users to compute this on a client (inheriting this lib)
* Do "out of range" values (0-65, 122-255) sort lexicographicly?
* I'm not sure what `Bias` is

I wonder if there's some room for simplification here. Can the spec be succinctly described? Can semver components be generalized such that `Position = [Component(16), Component(32), Component(64)]` or some form of generalized expansion? Can it be possible to write an insertion query without having to use this library, either through the client API or by a simplified spec that can be trivially reproduced?

**Comment by jsantell** [comment]
FWIW all the code looks good and covers any invariants I can think of

**Comment by Gozala** [comment]
> * What are the benefits for defining custom encoding ranges?

At the moment it mostly tried to make positions stored as attributes appear as valid property names.

In retrospect this may have not being a very important and perhaps it might be worth rethinking it in the futer.

> * Why is Base52 used for Major and Base62 used for Minor/Patch?

To map to a valid property names, we could have used even base 98, but  it did not feel like it would provide significant space saving but would make it difficult to use positions as dom attribute names.

> * Ultimately as a user-provided value, I question the effectiveness of "recommended" values to use: values should either be supported, or not, and a case where we're requiring end-users to compute this on a client (inheriting this lib)
> * Do "out of range" values (0-65, 122-255) sort lexicographicly?

They do

> * I'm not sure what `Bias` is

Bias provides slight position shift so that two concurrent inserts in between same exact items would end up in different positions. It's basically a hash / entity identifier of the item.

> 
> I wonder if there's some room for simplification here. Can the spec be succinctly described? Can semver components be generalized such that `Position = [Component(16), Component(32), Component(64)]` or some form of generalized expansion? Can it be possible to write an insertion query without having to use this library, either through the client API or by a simplified spec that can be trivially reproduced?

I think client should probably hide this from it's consumer. As of simplification, yeah it is certainly possible. I have since realized that representing position as `{row: varint, column: varint}` might have being simpler yet sufficient.

I found this article that takes slightly different, but seemingly simpler approach for [representing arbitrary sized decimals](https://terminusdb.com/blog/terminusdb-internals-3/#elementor-toc__heading-anchor-4) that preserves lexicographic ordering.

Unfortunately it is not clear how could they be mapped into string representation directly, but it is worth exploring in the future.

File: src/position/major.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ * and `B:98` denote 3 digit `Minor`. Each subsequent `Major` denotes increasingly
+ * large range for `Minor` position, making `A` and `z` a 27 digit ranges.
```
Comment (Gozala) [inline src/position/major.js:32]:
```suggestion
 * Major codes like `Z:65` and `a:97` denote two digit `Minor` range, codes `Y:66`
 * and `b:98` denote 3 digit `Minor`. Each subsequent `Major` denotes increasingly
 * large range for `Minor` position, making `A` and `z` a 27 digit ranges.
```

File: src/position/major.js (hunk 1 above)
Comment (Gozala) [reply src/position/major.js:30]: > These codes _do not_ line up with their ASCII representations, IIUC -- `A:65`, not `z`, no? `b:98` and `B:66` as well.

You're right thanks for caching these. Fix in the suggestion



> IIUC what's the difference between `A:65` and `a:97` if both signify minor patch size, this seems to have a lot of opinions on underlying encoding if "A" and "a" have some significant relationship other than its relative position

Yeah we treat [A..Z] as a top segment (for insertions in the top of the list) and [a..z] range as bottom segment (for insertions in the bottom of the list). Both ranges use shorter sizes and increasingly long ones towards the edges. This way key size increase logarithmically

**Pull request merged**